libcloud/test/test_connection.py

```diff
@@ -278,5 +280,49 @@ def test_log_curl(self):
         cmd = con._log_curl(method='HEAD', url=url, body=body, headers=headers)
         self.assertEqual(cmd, 'curl -i --head --compress http://example.com:80/test/path')
 
+    def _raise_socket_error(self):
+        raise socket.gaierror('')
+
+    @retry(sleep=2)
+    def _retry_test_with_sleep(self):
+        return self._raise_socket_error()
+
+    @retry(backoff=2)
+    def _retry_test_with_backoff(self):
+        return self._raise_socket_error()
+
+    @retry(timeout=2)
+    def _retry_test_with_timeout(self):
+        return self._raise_socket_error()
+
+    def test_retry_connection(self):
+        con = Connection()
+        con.connection = Mock()
+        con.connection.request = Mock(side_effect=socket.gaierror)
+
+        self.assertRaises(socket.gaierror, con.request, '/')
+
+    def test_retry_with_sleep(self):
+        try:
+            self._retry_test_with_sleep()
```

Comment:
These tests have the same problem.  They are only testing that the exception was thrown.  You need to verify that the underlying method was called multiple times.  There are a few ways to do this:

1. You can use mock to get the number of times a mocked method was called.  
2. You can patch it to throw the exception the first time (or first few times), then stop the patch at some point and verify that it returns successfully without throwing the exception.  

I'd also set the timeout and sleep values pretty low so it doesn't add a long amount of time to the unit tests.  You can sleep for fractional seconds.  timeout=1, sleep=0.1